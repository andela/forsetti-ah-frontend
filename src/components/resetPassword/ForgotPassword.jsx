import React from 'react';
import {
  Modal,
  ModalBody,
  InputGroup,
  InputGroupAddon,
  InputGroupText,
  Input,
  Spinner
} from 'reactstrap';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import ResetCancelIcon from '../common/ResetCancelIcon';
import { forgotPassword, openModalAction } from '../../actions';
import Button from '../common/Button';
import FormErrorText from '../common/FormErrorText';
import ToastMessage from '../common/ToastMessage';

class ForgotPasswordComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      emailError: '',
      isEmailInvalid: false
    };
  }

  /**
   * Handle form submission
   */
  onSubmit = async () => {
    const { email, isEmailInvalid } = this.state;
    const emailEmptyRegex = /^$/.test(email);
    const { dispatch } = this.props;
    const data = {
      email,
    };
    if (emailEmptyRegex || isEmailInvalid) {
      this.setState({
        emailError: 'Invalid email input',
      });
    } else {
      await dispatch(forgotPassword(data));
      const { status, message } = this.props;
      if (status === 200) {
        toast(<ToastMessage message={`${message}`} />, {
          type: 'success',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 0,
        });
      } else if (status === 404) {
        toast(<ToastMessage message={message} />, {
          type: 'error',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 5000,
        });
      } else {
        toast(<ToastMessage message='Email cannot be sent to the email provided' />, {
          type: 'error',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 5000,
        });
      }
    }
  }

  /**
   * Handle Input field change
   */
  handleChange = (e) => {
    this.setState({ [e.target.id]: e.target.value });
  }

  /**
   * Handle modal closing
   */
  closeModal = () => {
    const { dispatch } = this.props;
    dispatch(openModalAction());
  }

  /**
   * Clear input field on success
   */
  clearField = () => {
    this.setState({ email: '' });
  }

  /**
   * Check for invalid email input
   */
  checkEmailInput = (e) => {
    const validEmailRegex = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/.test(e.target.value);
    if (!validEmailRegex) {
      this.setState({
        emailError: 'Email is invalid',
        isEmailInvalid: true
      });
    } else {
      this.setState({
        emailError: '',
        isEmailInvalid: false
      });
    }
  }

  render() {
    const { email, emailError } = this.state;
    const { submit, isOpen } = this.props;
    return (
      <div>
        <Modal isOpen={isOpen} className='reset-modal'>
          <ResetCancelIcon clicked={this.closeModal} />
          <ModalBody>
            <div className='reset-form'>
              <h4 className='reset-form-header-text'>Forgot your password?</h4>
              <p className='reset-form-follower-text'>
                No hassles! Just type in the email address
                you used to register below.
              </p>
              <InputGroup className='reset-form-input'>
                <InputGroupAddon addonType='prepend' className='d-none d-md-flex'>
                  <InputGroupText className='reset-form-placeholder-text'>EMAIL</InputGroupText>
                </InputGroupAddon>
                <Input
                  id='email'
                  value={email}
                  onChange={this.handleChange}
                  onKeyUp={this.checkEmailInput}
                />
              </InputGroup>
              <FormErrorText message={emailError} className='form-error-text form-error-text-email' />
              <Button
                className='reset-form-btn'
                onClick={this.onSubmit}
              >
                  Send Link
                  &nbsp;
                { submit && <Spinner style={{ width: '1.5rem', height: '1.5rem' }} />}
              </Button>
            </div>
          </ModalBody>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  isModalOpen: state.reset.isModalOpen,
  submit: state.reset.submit,
  message: state.reset.message,
  status: state.reset.status
});

const ForgotPassword = connect(mapStateToProps)(ForgotPasswordComponent);
export { ForgotPassword, ForgotPasswordComponent };
