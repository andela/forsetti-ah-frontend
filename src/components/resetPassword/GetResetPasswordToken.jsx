import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getToken } from '../../actions';

class GetResetPasswordToken extends Component {
  /**
   * Get the user token from url
   */
  getToken = () => {
    const { location, dispatch, history } = this.props;
    const token = location.search.split('=')[1];
    dispatch(getToken(token));
    history.push('/');
  }

  render() {
    return (
      <div>
        {this.getToken()}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  token: state.reset.token
});

export default connect(mapStateToProps)(GetResetPasswordToken);
