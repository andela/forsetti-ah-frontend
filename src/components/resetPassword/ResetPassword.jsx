import React from 'react';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import {
  Modal,
  ModalBody,
  InputGroup,
  InputGroupAddon,
  InputGroupText,
  Input,
  Spinner
} from 'reactstrap';
import Button from '../common/Button';
import FormErrorText from '../common/FormErrorText';
import ResetCancelIcon from '../common/ResetCancelIcon';
import { resetPassword, openModalAction, closeResetModal } from '../../actions';
import ToastMessage from '../common/ToastMessage';

class ResetPasswordComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      password: '',
      confirmPassword: '',
      passwordError: '',
      confirmPasswordError: '',
      isPasswordInvalid: false,
      doPasswordsMatch: false
    };
  }

  /**
   * Handle form submission
   */
  onSubmit = async () => {
    const { dispatch, history, token } = this.props;
    const { password, isPasswordInvalid, doPasswordsMatch } = this.state;
    const passwordEmptyRegex = /^$/.test(password);
    const data = {
      password
    };
    if (passwordEmptyRegex || isPasswordInvalid || doPasswordsMatch) {
      if (doPasswordsMatch) {
        this.setState({
          confirmPasswordError: 'Password does not match'
        });
      } else {
        this.setState({
          passwordError: 'Invalid input',
          confirmPasswordError: 'Invalid input'
        });
      }
    } else {
      await dispatch(resetPassword(data, token));
      const { status, message } = this.props;
      if (status === 201) {
        toast(<ToastMessage message={`${message}. Redirecting...`} />, {
          type: 'success',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 0,
        });
        setTimeout(() => {
          dispatch(closeResetModal());
          dispatch(openModalAction());
        }, 5000);
      } else if (status === 409) {
        toast(<ToastMessage message={`${message}`} />, {
          type: 'error',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 0,
        });
      } else if (status === 401) {
        toast(<ToastMessage message='Token expired. Please request for a change again' />, {
          type: 'error',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 0,
        });
      } else if (status === 422) {
        toast(<ToastMessage message={`${message}`} />, {
          type: 'error',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 0,
        });
      } else {
        toast(<ToastMessage message='Password cannot be changed at the moment. Please request for a change again' />, {
          type: 'error',
          closeButton: false,
          hideProgressBar: true,
          onClose: () => this.clearField(),
          autoClose: 0,
        });
      }
    }
  }

  /**
   * Handle Input field change
   */
  handleChange = (e) => {
    this.setState({ [e.target.id]: e.target.value });
  }

  /**
   * Clear input field on success
   */
  clearField = () => {
    this.setState({ password: '' });
  }

  /**
   * Hande password input check
   */
  checkPasswordInput = (e) => {
    const validPasswordRegex = /^(?=.*[a-zA-Z])(?=.*[0-9]){8}/.test(e.target.value);
    if (!validPasswordRegex) {
      this.setState({
        passwordError: 'Should be alphanumeric and atleast 8 characters long',
        isPasswordInvalid: true
      });
    } else {
      this.setState({
        passwordError: '',
        isPasswordInvalid: false
      });
    }
  }

  /**
   * Hande password match check
   */
  checkPasswordMatch = (e) => {
    const { password } = this.state;
    if (password !== e.target.value) {
      this.setState({
        confirmPasswordError: 'Passwords don\'t match',
        doPasswordsMatch: true
      });
    } else {
      this.setState({
        confirmPasswordError: '',
        doPasswordsMatch: false
      });
    }
  }

  /**
   * Handle modal closing
   */
  closeModal = () => {
    const { dispatch } = this.props;
    dispatch(openModalAction());
    dispatch(closeResetModal());
  }

  render() {
    const {
      password, passwordError, confirmPassword, confirmPasswordError
    } = this.state;
    const { submit, isOpen } = this.props;
    return (
      <div>
        <Modal isOpen={isOpen} className='reset-modal'>
          <ResetCancelIcon clicked={this.closeModal} />
          <ModalBody>
            <div className='reset-form'>
              <h4 className='reset-form-header-text reset-text'>Reset Password</h4>
              <InputGroup className='reset-form-input'>
                <InputGroupAddon addonType='prepend' className='d-none d-md-flex'>
                  <InputGroupText className='reset-form-placeholder-text'>PASSWORD</InputGroupText>
                </InputGroupAddon>
                <Input
                  type='password'
                  id='password'
                  value={password}
                  onChange={this.handleChange}
                  onKeyUp={this.checkPasswordInput}
                />
              </InputGroup>
              <FormErrorText message={passwordError} className='form-error-text form-error-text-email' />

              <InputGroup className='reset-form-input'>
                <InputGroupAddon addonType='prepend' className='d-none d-md-flex'>
                  <InputGroupText className='reset-form-placeholder-text'>CONFIRM</InputGroupText>
                </InputGroupAddon>
                <Input
                  type='password'
                  id='confirmPassword'
                  value={confirmPassword}
                  onChange={this.handleChange}
                  onKeyUp={this.checkPasswordMatch}
                />
              </InputGroup>
              <FormErrorText message={confirmPasswordError} className='form-error-text form-error-text-email' />
              <Button
                className='reset-form-btn'
                onClick={this.onSubmit}
              >
                Reset
                &nbsp;
                { submit && <Spinner style={{ width: '1.5rem', height: '1.5rem' }} />}
              </Button>
            </div>
          </ModalBody>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  submit: state.reset.submit,
  message: state.reset.message,
  status: state.reset.status,
  token: state.reset.token
});

const ResetPassword = connect(mapStateToProps)(ResetPasswordComponent);

export { ResetPassword, ResetPasswordComponent };
