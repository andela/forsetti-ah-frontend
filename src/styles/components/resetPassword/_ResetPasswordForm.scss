.reset-form {

  &-input {
    margin: 2rem 0;
  }

  &-placeholder-text {
    background-color: $primary-color;
    color: $white-color;
    font-size: .7rem;
    width: 7rem;
  }


  &-header-text {
    color: $primary-color;
    margin-bottom: 2rem;
    text-align: center;
    text-transform: uppercase;
  }

  h4.reset-form-header-text.reset-text {
    margin-top: 3rem;
  }

  &-follower-text {
    color: $primary-color;
    margin-bottom: 2rem;
    text-align: center;
  }

  &-btn {
    width: 9rem;
    display: block;
    margin: 3rem auto;
    text-transform: uppercase;
    background-color: $primary-color;
    color: $white-color;

    @media screen and (max-width:425px) {
      width: 100%;
      margin: 3rem auto;
    }

    &:hover {
      background-color: $secondary-color;
    }
  }

  @media screen and (max-width:425px) {
    font-size: 1rem;
  }

  @media screen and (max-width: 575px) {
    width: 15rem;
    margin: 5rem auto;
  }
}
