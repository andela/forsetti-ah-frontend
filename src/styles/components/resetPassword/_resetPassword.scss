.reset-modal {
  max-width: 45rem;
  min-width: 15rem;
  margin-top: 7rem;
  overflow-x: scroll;
 }
