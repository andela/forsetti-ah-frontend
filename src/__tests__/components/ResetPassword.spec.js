import React from 'react';
import { shallow } from 'enzyme';
import { ResetPasswordComponent } from '../../components/resetPassword/ResetPassword';
import { Spinner } from 'reactstrap';

describe('<ResetPasswordComponent />', () => {
  let component;
  beforeAll(() => {
    const props = {
      reset: {
        submit: false,
        message: '',
        status: ''
      }
    }
    component = shallow(<ResetPasswordComponent {...props} />);
  });

  it('correctly sets state for password input', () => {
    const input = component.find('Input').at(0);
    input.value = 'Password2333';
    const mockedInputEvent = { target: {
      id: 'password',
      value: 'Password2333'
    }};
    input.simulate('change', mockedInputEvent);
    expect(component.state().password).toEqual('Password2333');
  });

  it('correctly sets state when password is invalid', () => {
    const input = component.find('Input').at(0);
    input.value = 'Password2333';
    const mockedInputEvent = { target: {
      id: 'password',
      value: 'Passwo'
    }};
    input.simulate('keyup', mockedInputEvent);
    expect(component.state().passwordError).toEqual('Should be alphanumeric and atleast 8 characters long');
  });

  it('correctly sets state when password is valid', () => {
    const input = component.find('Input').at(0);
    input.value = 'Password2333';
    const mockedInputEvent = { target: {
      id: 'password',
      value: 'Password2333'
    }};
    input.simulate('keyup', mockedInputEvent);
    expect(component.state().passwordError).toEqual('');
  });

  it('should not render the spinner when rendering submit is false', () => {
    const props = {
      auth: {
        submit: false,
      }
    };
    component = shallow(<ResetPasswordComponent {...props} />);
    expect(component.find(Spinner)).toHaveLength(0);
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
