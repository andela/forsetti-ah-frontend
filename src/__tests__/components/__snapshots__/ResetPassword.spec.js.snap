// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPasswordComponent /> renders correctly 1`] = `ShallowWrapper {}`;
